Monika After Story/game/zz_apikeys.rpy

```diff
@@ -0,0 +1,556 @@
+
+        IN:
+            dirty_key - key to clean
+
+        RETURNS: cleaned key
+        """
+        return dirty_key.replace("\n", "").replace("\r", "")
+
+
+    def _run_on_change(feature, api_key):
+        """
+        Runs on change for a feature with api key
+
+        IN:
+            feature - feature to run on change for
+            api_key - api key to run on change with
+
+        RETURNS: tuple of the following format:
+            [0] - True if valid key, False if not
+            [1] - error message to show
+        """
+        on_change = Data.registered_api_keys[feature][1]
+        if on_change is not None:
+            try:
+                rv = on_change(api_key)
+
+                if not isinstance(rv, tuple) or len(rv) < 2:
+                    return True, ""
+
+                return rv
+
+            except Exception as e:
+                mas_utils.mas_log.error(
+                    "crash when running on_change for feature {0} - {1}".format(
+                        feature,
+                        repr(e)
+                    )
+                )
+
+                return False, "on-change crash - see logs"
+
+        return True, ""
+
+
+    def screen_clear(feature):
+        """
+        Called when the clear button is used on the api key screen
+
+        IN:
+            feature - the feature whose key is being cleared
+        """
+        if not feature_registered(feature) or not store.mas_hasAPIKey(feature):
+            return
+
+        # clear key
+        Data.api_keys.pop(feature)
+        save_keys()
+
+        # on change
+        _run_on_change(feature, "")
+
+
+    def screen_paste(feature):
+        """
+        Called when the paste button is used on the api key screen.
+
+        IN:
+            feature - the feature whose key is being pasted
+        """
+        if not feature_registered(feature):
+            return
+
+        # grab key 
+        new_key = pygame.scrap.get(pygame.SCRAP_TEXT).strip()
+        if not new_key:
+            # null key is not counted
+            return
+
+        # clear newlines
+        new_key = clean_key(new_key)
+
+        # on change
+        key_valid, err_msg = _run_on_change(feature, new_key)
+
+        if key_valid:
+            # set key
+            Data.api_keys[feature] = new_key
+            save_keys()
+
+        else:
+            # show message box
+            store.renpy.show_screen(
+                "dialog",
+                message=err_msg,
+                ok_action=store.Hide("dialog")
+            )
+
+
+    def screen_update_cert():
+        """
+        Called when the update cert button is used on the api key screen
+        """
+        if not store.mas_can_import.certifi():
+            # no certifi enabled - the button shouldn't be visislbe
+            return
+
+        screen_data = MASUpdateCertScreenData()
+        screen_data.start()
+
+        store.renpy.show_screen("mas_update_cert", screen_data)
```

Comment:
Screen prediction is for performance, right? I doubt the cert update screen will be shown all that often where the perf gain is useful.

Reply:
It's just I think this could be done via screens. You're doing `action Function(store.mas_api_keys.screen_update_cert)` anyway, so we could use `ShowScreen` for that.

You are right, it's not a big deal for this case, though.